### PR TITLE
Use separate depth than finality depth for merge set calculations after HF

### PR DESCRIPTION
**Pull request by someone235**

### CODE REVIEW

File: app/protocol/flows/v5/blockrelay/ibd.go [hunk 1]
```diff
@@ -268,6 +277,42 @@ func (flow *handleIBDFlow) negotiateMissingSyncerChainSegment() (*externalapi.Do
 	return syncerHeaderSelectedTipHash, highestKnownSyncerChainHash, nil
 }
 
+func (flow *handleIBDFlow) isViolatingBoundedMergeDepth(highestKnownSyncerChainHash *externalapi.DomainHash, relayBlock *externalapi.DomainBlock) (bool, error) {
+	// Note: we don't wait until the hard fork is activated before applying these rules in the p2p leve.
```
Comment (michaelsutton) [inline app/protocol/flows/v5/blockrelay/ibd.go:281]: leve -> level

File: app/protocol/flows/v5/blockrelay/ibd.go [hunk 2]
```diff
@@ -268,6 +277,42 @@ func (flow *handleIBDFlow) negotiateMissingSyncerChainSegment() (*externalapi.Do
...
+	virtualSelectedParentHeader, err := flow.Domain().Consensus().GetBlockHeader(virtualSelectedParent)
+	if err != nil {
+		return false, err
+	}
+
+	if virtualSelectedParentHeader.BlueWork().Cmp(relayBlock.Header.BlueWork()) <= 0 {
+		return false, nil
+	}
+
+	highestSharedChainBlock, found, err := flow.Domain().Consensus().HighestChainBlockInPastOf(highestKnownSyncerChainHash)
```
Comment (michaelsutton) [inline app/protocol/flows/v5/blockrelay/ibd.go:298]: `highestSharedChainBlock` sounds like it is the chain block on both sides. Perhaps rename to `highestChainBlockInSyncerPast`

File: app/protocol/flows/v5/blockrelay/ibd.go [hunk 3]
```diff
@@ -268,6 +277,42 @@ func (flow *handleIBDFlow) negotiateMissingSyncerChainSegment() (*externalapi.Do
 	return syncerHeaderSelectedTipHash, highestKnownSyncerChainHash, nil
 }
 
+func (flow *handleIBDFlow) isViolatingBoundedMergeDepth(highestKnownSyncerChainHash *externalapi.DomainHash, relayBlock *externalapi.DomainBlock) (bool, error) {
+	// Note: we don't wait until the hard fork is activated before applying these rules in the p2p leve.
+	// If such side chain is created before the hard fork, it will be downloaded if it'll be merged to
+	// the selected chain by another node.
+	virtualSelectedParent, err := flow.Domain().Consensus().GetVirtualSelectedParent()
```
Comment (michaelsutton) [inline app/protocol/flows/v5/blockrelay/ibd.go:284]: Isn't it better practice to extract `flow.Domain().Consensus()` to a local variable to use in all calls ?

File: domain/consensus/consensus.go [hunk 4]
```diff
@@ -881,3 +886,62 @@ func (s *consensus) IsChainBlock(blockHash *externalapi.DomainHash) (bool, error
...
+
+	headersSelectedTip, err := s.headersSelectedTipStore.HeadersSelectedTip(s.databaseContext, stagingArea)
+	if err != nil {
+		return nil, false, err
+	}
+
+	lowHash := pruningPoint
+	highHash := headersSelectedTip
+	for {
+		blockLocator, err := s.syncManager.CreateHeadersSelectedChainBlockLocator(stagingArea, lowHash, highHash)
```
Comment (michaelsutton) [inline domain/consensus/consensus.go:917]: The code might be simpler if we directly implement a binary search here w/o going through block locator logic (also makes it `log(n)` over `(log(n))^2`). Though I can see why reusing has benefits.

**Review by michaelsutton — CHANGES_REQUESTED** [review]

File: domain/consensus/processes/mergedepthmanager/merge_depth_manager.go [hunk 5]
```diff
@@ -75,49 +71,67 @@ func (mdm *mergeDepthManager) CheckBoundedMergeDepth(stagingArea *model.StagingA
...
 
 	nonBoundedMergeDepthViolatingBlues := make([]*externalapi.DomainHash, 0, len(ghostdagData.MergeSetBlues()))
 
-	finalityPoint, err := mdm.finalityManager.FinalityPoint(stagingArea, blockHash, isBlockWithTrustedData)
 	if err != nil {
 		return nil, err
 	}
 	for _, blue := range ghostdagData.MergeSetBlues() {
-		notViolatingFinality, err := mdm.dagTopologyManager.IsInSelectedParentChainOf(stagingArea, finalityPoint, blue)
+		notViolatingFinality, err := mdm.dagTopologyManager.IsInSelectedParentChainOf(stagingArea, mergeDepthRoot, blue)
```
Comment (someone235) [inline domain/consensus/processes/mergedepthmanager/merge_depth_manager.go:134]: Rename this var

**Review by michaelsutton — APPROVED** [review]

**Review by michaelsutton — APPROVED** [review]

**Pull request merged**